### PR TITLE
Queues - Elizabeth Deutsch - Ride-Share-Two

### PR DESCRIPTION
**Pull request by edeutschie**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |   One design decision for me was whether or not to use an array or a hash to access the CSV data in the initialize methods.  I chose to go with an array because I had familiarity with that format from the Bank Accounts project.  Arrays seemed to work fine, but if I had time, I would like to refactor it to use hashes so that I could have a better understanding of how to use hashes in a scenario like this.     |
| Describe a concept that you gained more clarity on as you worked on this assignment. | I gained more clarity on how methods can be called across classes and how drawing diagrams like the farmer's market diagram we worked with in class can help me visualize those connections and plan how to write them in code form.   |
| Describe a _nominal test_ that you wrote for this assignment. |  A nominal test I wrote for the Driver class for the find_driver method was to look for a driver with an ID I knew existed and test to make sure an instance of the Driver class was returned when the method was called.    |
| Describe an _edge case test_ that you wrote for this assignment. |  I wrote some edge case tests in the Driver class for the find_driver method.  There were 100 drivers in the CSV file.  I wrote one test to look for the driver with driver_id 100 (to make sure it returned an instance of the Driver class) and one to look for the driver with driver_id 101 (to make sure it returned "not a driver").  |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? |  I am much better at writing tests and then the code.  I still need to work on writing pseudocode first.  It doesn't come as automatically.  |

### CODE REVIEW

File: specs/driver_spec.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+end
+
+# test for "list_driver_trips method"? spoke with a tutor and he said you trust the tests in the Trip class to cover this.
+
+describe "average_rating" do
+  let(:driver) {RideShareTwo::Driver.find_driver(21)}
+
+  it "retrieves an average rating for a driver based on all trips taken" do
+    driver.average_rating.must_equal 30.0 / 11.0
+  end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:41]: What happens if the driver hasn't taken any trips yet?

File: specs/rider_spec.rb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+  it "retrieves the list of all previous driver instances & returns as an array" do
+    rider_trips = rider_object.list_drivers_for_rider
+    rider_trips[0].must_be_instance_of RideShareTwo::Driver
+  end
+
+  it "returns an array of drivers without duplicates" do
+    rider_trips = rider_object.list_drivers_for_rider
+    rider_trips.length.must_equal 3
+  end
```
Comment (droberts-sea) [inline specs/rider_spec.rb:39]: As with `Driver.average_rating`, what happens if the rider has taken 0 trips?

File: specs/trip_spec.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+  it "Returns an array (of all trips instances for a given driver ID)" do
+    driver_trips_list
+    driver_trips_list.must_be_kind_of Array, "Oops - the result is not an array"
+  end
+
+  it "Returns an array of all trips instances" do
+    driver_trips_list
+    driver_trips_list[0].must_be_instance_of RideShareTwo::Trip, "Oops - the item in the array is not a trip instance."
+  end
```
Comment (droberts-sea) [inline specs/trip_spec.rb:80]: What if there are no trips associated with that driver?

File: specs/trip_spec.rb [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+  it "Returns an array (of all rider instances for a given rider ID)" do
+    rider_trips_list
+    rider_trips_list.must_be_kind_of Array, "Oops - the result is not an array"
+  end
+
+  it "Returns an array of all trips instances" do
+    rider_trips_list
+    rider_trips_list[0].must_be_instance_of RideShareTwo::Trip, "Oops - the item in the array is not a trip instance."
+  end
```
Comment (droberts-sea) [inline specs/trip_spec.rb:94]: What if there are no trips associated with that rider?

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

| Feature | Feedback    |
| :------------- | :------------- |
|  **Baseline** |   |
|   Used Git Regularly	|   yes	|
|   Answer comprehension questions	|   yes	|
|  **Driver** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   mostly - see inline comment |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips list to calculate the average rating | yes |
|  **Rider** |   |
|   Uses the `all` method in the `find` method | yes |
|   Has appropriate edge-case tests for each method in the class |   mostly - see inline |
|   Created a method that uses a method from the `Trip` object to retrieve the list of trips | yes |
|   Created a method that uses the internal trips method to retrieve the associated drivers | yes |
|  **Trip** |   |
|   Reads the CSV file in the `all` method | yes |
|   Has appropriate edge-case tests for each method in the class |   mostly - see inline |
|   Created a method that uses a method from the `Driver` to retrieve the associated driver instance  |   yes |
|   Created a method that uses a method from the `Rider` to retrieve the associated rider instance |   yes |
|   Created a method to retrieve all trips by driver id |    yes |
|   Created a method to retrieve all trips by rider id |     yes |
|   **Overall**	|   	|

Great job! Code is clear and readable, and git habits look solid. The one testing pattern I'm seeing you consistently miss is "what happens if this method ought not to find anything" - see inline comments for specific places. In general, keep up the good work!
